### PR TITLE
get rusty and layered to build for older kernel

### PR DESCRIPTION
**Pull request by likewhatevs**
opening pr to show changes to get rusty and layered to build for older kernel

### CODE REVIEW

File: .github/workflows/caching-build.yml [hunk 1]
```diff
@@ -14,8 +14,8 @@ jobs:
       - run: sudo apt update
       - run: sudo apt install -y git --no-install-recommends
       # get latest head commit of sched_ext for-next
-      - run: echo "SCHED_EXT_KERNEL_COMMIT=$(git ls-remote https://git.kernel.org/pub/scm/linux/kernel/git/tj/sched_ext.git heads/for-next | awk '{print $1}')" >> $GITHUB_ENV
 
+      - run: echo "SCHED_EXT_KERNEL_COMMIT=21f4c19" >> $GITHUB_ENV
```
Comment (hodgesds) [inline .github/workflows/caching-build.yml:16]: does this need updated on each update?

File: rust/scx_rustland_core/assets/bpf/main.bpf.c [hunk 2]
```diff
@@ -498,8 +498,14 @@ static void dispatch_task(struct task_struct *p, u64 dsq_id,
...
-	if (cpu != bpf_get_smp_processor_id())
-		scx_bpf_kick_cpu(cpu, SCX_KICK_IDLE);
+	if (cpu != bpf_get_smp_processor_id()){
+		if (sib >= 0 && (sib_cctx = lookup_cpu_ctx(sib)) &&
+			sib_cctx->maybe_idle) {
+			gstat_inc(GSTAT_EXCL_WAKEUP, cctx);
+			scx_bpf_kick_cpu(sib, 0);
+		}
+	}
+		// scx_bpf_kick_cpu(cpu, SCX_KICK_IDLE);
```
Comment (hodgesds) [inline rust/scx_rustland_core/assets/bpf/main.bpf.c:508]: delete?

File: rust/scx_utils/src/compat.rs [hunk 3]
```diff
@@ -218,7 +219,7 @@ macro_rules! scx_ops_open {
 macro_rules! scx_ops_load {
     ($skel: expr, $ops: ident, $uei: ident) => { 'block: {
         scx_utils::paste! {
-            scx_utils::uei_set_size!($skel, $ops, $uei);
+            // scx_utils::uei_set_size!($skel, $ops, $uei);
```
Comment (hodgesds) [inline rust/scx_utils/src/compat.rs]: For existing back ports I just skipped this macro and loaded the skels manually, but if this works.

File: scheds/rust/scx_layered/src/bpf/main.bpf.c [hunk 4]
```diff
@@ -1508,30 +1496,38 @@ void BPF_STRUCT_OPS(layered_dump, struct scx_dump_ctx *dctx)
...
+	if (tctx->layer >= 0 && tctx->layer < nr_layers)
+		__sync_fetch_and_add(&layers[tctx->layer].nr_tasks, -1);
 }
 
 s32 BPF_STRUCT_OPS_SLEEPABLE(layered_init)
 {
 	struct bpf_cpumask *cpumask;
 	int i, j, k, nr_online_cpus, ret;
 
-	ret = scx_bpf_create_dsq(HI_FALLBACK_DSQ, -1);
```
Comment (hodgesds) [inline scheds/rust/scx_layered/src/bpf/main.bpf.c:1528]: pretty sure these DSQs are needed

File: scheds/rust/scx_layered/src/bpf/main.bpf.c [hunk 5]
```diff
@@ -1623,21 +1619,6 @@ s32 BPF_STRUCT_OPS_SLEEPABLE(layered_init)
 				case MATCH_NICE_EQUALS:
 					dbg("%s NICE_EQUALS %d", header, match->nice);
 					break;
-				case MATCH_USER_ID_EQUALS:
```
Comment (hodgesds) [inline scheds/rust/scx_layered/src/bpf/main.bpf.c]: these should all be compatible with older kernels

File: .github/workflows/caching-build.yml (hunk 1 above)
Comment (likewhatevs) [reply .github/workflows/caching-build.yml:16]: yes for each kernel commit/branch/repo triple you want to backport against.

File: rust/scx_utils/src/compat.rs (hunk 3 above)
Comment (likewhatevs) [reply rust/scx_utils/src/compat.rs]: ended up doing the same

**Comment by likewhatevs** [comment]
I think rusty works also now (local stress test logs look roughly the same as those from the main repo's CI)?

File: .github/workflows/caching-build.yml (hunk 1 above)
Comment (likewhatevs) [reply .github/workflows/caching-build.yml:16]: ahh for clarity, those are params are on the kernel sources, not the library repo. the library repo commit used is that which the backport branch is rebased upon (i.e. this stacks on that).

**Pull request merged**